peer/connector.go

```diff
@@ -79,56 +83,100 @@ func (conn *connector) initialise(privateKey []byte, publicKey []byte, connect [
 	}
 	conn.staticClients = make([]*upstream.Upstream, staticCount)
 
-	// error code for goto fail
-	errX := error(nil)
-
 	// initially connect all static sockets
+	wg := sync.WaitGroup{}
+	errCh := make(chan error, len(connect))
+
+	conn.log.Debugf("static connection count: %d", len(connect))
+
 	for i, c := range connect {
-		address, err := util.NewConnection(c.Address)
-		if nil != err {
-			log.Errorf("client[%d]=address: %q  error: %s", i, c.Address, err)
-			errX = err
-			goto fail
-		}
-		serverPublicKey, err := hex.DecodeString(c.PublicKey)
-		if nil != err {
-			log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
-			errX = err
-			goto fail
-		}
+		wg.Add(1)
 
-		// prevent connection to self
-		if bytes.Equal(publicKey, serverPublicKey) {
-			errX = fault.ErrConnectingToSelfForbidden
-			log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, errX)
-			goto fail
-		}
+		// start new goroutine for each connection
+		go func(conn *connector, c Connection, i int, wg *sync.WaitGroup, ch chan error) {
```

Comment:
Change from loop into goroutine, I guess intention is to provide faster initialization time, and `lock` is only used during connector initialization, I am curious about how much faster in this way?

Reply:
I think it's much faster, especially in case that the network is slow. If the connection is initialized sequently, the operation would be blocked.